### ISSUE TITLE
AST as array

### ISSUE THREAD

**Issue by vladar** [webonyx/graphql-php#108]
This is required to leverage [libgraphqlparser](https://github.com/graphql/libgraphqlparser) in future.

**Comment by joseluisq** [comment]
:+1:

**Comment by vladar** [comment]
Unfortunately, after looking deeper into this I realized that this would be a really major break + it will make dealing with ASTs way less pleasant from IDE. 

I implemented a different solution to enable compatibility with libgraphqlparser: now it is possible to create AST from array returned by libgraphqlparser. 

To make such conversion performant I had to replace all arrays of nodes (such as `selections`, `directives`, `arguments`, etc) with `NodeList` class which lazily converts arrays from libgraphqlparser to Node instances.

This way typical integration with libgraphqlparser (e.g. using [graphql-parser-php](https://github.com/dosten/graphql-parser-php)) would looks like this:

```php
$parser = new \GraphQL\Parser(); // graphql-parser-php
$astAsArray = $parser->parse('query { name }');

// Convert AST array to lazily-evaluated DocumentNode:
$ast = GraphQL\Language\AST\DocumentNode::fromArray($astAsArray);
$result = GraphQL\GraphQL::execute($schema, $ast);
```

**Comment by vladar** [comment]
Released in v0.10.0. See [docs](http://webonyx.github.io/graphql-php/reference/#graphqlutilsast) on how to convert AST from/to array. 

This is a lazy operation (array is converted to AST as it is traversed), so should be performant enough.
